File: fee_crawler/engine/golden.py
"""Golden-set regression — catch extraction/pipeline regressions before publish.

Compares the latest verified fees for each golden institution against a
hand-curated expected snapshot. A diff (missing key, wrong amount beyond
tolerance, or an unexpected extra key) is a regression. Runs as a gate before
the national roll-up publishes; failures alert and can block the swap.
"""

from __future__ import annotations

from dataclasses import dataclass, field

import asyncpg


@dataclass
class GoldenDiff:
    crawl_target_id: int
    missing: list[str] = field(default_factory=list)          # expected but absent
    wrong_amount: list[tuple[str, float, float]] = field(default_factory=list)  # (key, expected, got)
    unexpected: list[str] = field(default_factory=list)       # present but not expected

    @property
    def clean(self) -> bool:
        return not (self.missing or self.wrong_amount or self.unexpected)


async def check_golden_set(pool: asyncpg.Pool) -> list[GoldenDiff]:
    """Return a diff per golden institution. Empty list => nothing configured."""
    async with pool.acquire() as conn:
        targets = await conn.fetch("SELECT crawl_target_id FROM golden_institutions")
        diffs: list[GoldenDiff] = []
        for t in targets:
            tid = t["crawl_target_id"]
            expected = {
                r["canonical_fee_key"]: (r["expected_amount"], r["tolerance"])
                for r in await conn.fetch(
                    "SELECT canonical_fee_key, expected_amount, tolerance "
                    "FROM golden_fees WHERE crawl_target_id=$1",
                    tid,
                )
            }
            # Latest verified value per canonical key for this institution.
            got = {
                r["canonical_fee_key"]: r["amount"]
                for r in await conn.fetch(
                    """
                    SELECT DISTINCT ON (canonical_fee_key) canonical_fee_key, amount
                      FROM fees_verified WHERE institution_id=$1
                     ORDER BY canonical_fee_key, created_at DESC
                    """,
                    tid,
                )
            }
            diff = GoldenDiff(tid)
            for key, (exp_amt, tol) in expected.items():
                if key not in got:
                    diff.missing.append(key)
                elif exp_amt is not None and got[key] is not None:
                    if abs(float(got[key]) - float(exp_amt)) > float(tol):
                        diff.wrong_amount.append((key, float(exp_amt), float(got[key])))
            for key in got:
                if key not in expected:
                    diff.unexpected.append(key)
            diffs.append(diff)
    return diffs


async def golden_regressions(pool: asyncpg.Pool) -> list[GoldenDiff]:
    """Only the golden institutions that regressed (non-clean diffs)."""
    return [d for d in await check_golden_set(pool) if not d.clean]

File: src/lib/engine-db/golden.ts
/**
 * Golden-set status for the console — pass/fail per pinned institution, the
 * accuracy gate that blocks Atlas from publishing a regression.
 */

import { sql } from "@/lib/crawler-db/connection";

export interface GoldenStatus {
  total: number;
  regressions: number;
  clean: boolean;
}

export async function getGoldenStatus(): Promise<GoldenStatus> {
  try {
    // A golden institution "regresses" if any expected fee is missing or an
    // amount is off beyond tolerance vs the latest verified value.
    const rows = await sql<
      { crawl_target_id: string; expected: string; matched: string }[]
    >`
      WITH latest AS (
        SELECT DISTINCT ON (institution_id, canonical_fee_key)
               institution_id, canonical_fee_key, amount
          FROM fees_verified ORDER BY institution_id, canonical_fee_key, created_at DESC
      )
      SELECT g.crawl_target_id,
             count(*)                                                          AS expected,
             count(*) FILTER (
               WHERE l.canonical_fee_key IS NOT NULL
                 AND (g.expected_amount IS NULL OR l.amount IS NULL
                      OR abs(l.amount - g.expected_amount) <= g.tolerance))    AS matched
        FROM golden_fees g
        LEFT JOIN latest l ON l.institution_id = g.crawl_target_id
                          AND l.canonical_fee_key = g.canonical_fee_key
       GROUP BY g.crawl_target_id
    `;
    const total = rows.length;
    const regressions = rows.filter((r) => Number(r.matched) < Number(r.expected)).length;
    return { total, regressions, clean: regressions === 0 };
  } catch {
    return { total: 0, regressions: 0, clean: true };
  }
}
